### DISCUSSION TITLE
How to make user oriented bot?

### DISCUSSION THREAD

**Discussion by arpitgoyall** [python-telegram-bot/python-telegram-bot#2655] [Q&A] [answered]
I want to make a telegram bot that should work differently for each user differently according to their settings, for example replying to the chat is 'on' for all the users by default but how can give the option to choose a user whether he wants to keep on or off it. I am very confused about how can I designed that system and also I want the 'login with the telegram' feature when a new user comes, these 2 questioned really threatened me. Please help if anyone has an answer. I have seen many bots those reply with an inline keyboard when you run /settings command and you can customize that bot according to you.

**Comment by harshil21** [comment, marked as answer]
If you want to present a user with keyboard buttons for settings, see [inlinekeyboard.py](https://github.com/python-telegram-bot/python-telegram-bot/blob/master/examples/inlinekeyboard.py)

Ideally you'll want an database or persistence to store those settings. 

For databases, you can have a look at [ptbcontrib/postgres_persistence](https://github.com/python-telegram-bot/ptbcontrib/tree/main/ptbcontrib/postgres_persistence).

Or you can see our in built PicklePersistence ([example](https://github.com/python-telegram-bot/python-telegram-bot/blob/master/examples/persistentconversationbot.py)). 

I don't understand the 'login with telegram' feature since the user is already talking to the bot inside telegram. That is only useful when you are logging into telegram from the web I think.

**Reply by arpitgoyall** [reply]
Sorry for replying late, I was checking how it works and I got the point of the database working but I think you are confused in 'login with telegram' my mean was the authentication of a user before using bot like in telegram chat forwarding bots, first, we need to login into them by phone no. and just by OTP. The same feature I want to use but don't know-how. I need to store API id, API hash, and session key. The final goal is to use their telegram account for chat forwarding. Thanks for the help!

**Reply by harshil21** [reply]
Oh alright. If you want the user's phone number, you can request it by following [this code snippet ](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Code-snippets#requesting-location-and-contact-from-user).

However, if you want to store their API ID, hash, session key, and also use their account for chat forwarding, then you'll have to use a [userbot](https://telegra.ph/How-a-Userbot-superacharges-your-Telegram-Bot-07-09).

**Reply by arpitgoyall** [reply]
Thank you so much this is helping me a lot, my telegram bot is almost ready and I am publishing it on Heroku but its database file named 'conversationbot' getting deleted when the dyno restarts, what to do now?

**Reply by harshil21** [reply]
That is because Heroku uses an [ephemeral storage system](https://devcenter.heroku.com/articles/dynos#ephemeral-filesystem). You can use their [Postgres addon](https://elements.heroku.com/addons/heroku-postgresql) to save instead.

**Reply by arpitgoyall** [reply]
Yeah i saw that after installing it i got the database URL but what to do next sorry but i don't know much about databases. At first you gave me two options of using databases i choosed easy in-built one which is working fine on local machine and other one(ptcontrib) is giving some code to deploy it and Heroku Postgres plugin have different code which one i have to use and how to access it after deploying, I am very confused how it works, **please tell me easy way so that i can store user data and access it on heroku**?

Thanks again for reply!

**Reply by harshil21** [reply]
imo, it's better to use Heroku's postgres addon since you're using heroku to deply. You can find their documentation [here](https://devcenter.heroku.com/articles/heroku-postgresql#connecting-in-python). Also you can ask more about heroku in the [off-topic group](https://t.me/pythontelegrambottalk) so you'll get answers quicker.
